rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,654 @@
+
+Cluster admins may view the migration history via `tctl migrations ls`.
+
+### Option 1: Stricter Resource Versioning
+
+Without a more concrete strategy for resource versioning, it is impossible to have different versions of Auth running
+concurrently. Auth needs to be aware of the exact version of a resource that clients are requesting to determine if the
+version of said resource stored in the backend is at the same version, if the stored version is capable of being
+downgraded to that version, if the stored version is capable of being updated to the requested version, or whether the
+request cannot be honored due to version incompatibility.
+
+The only backend operation that uses a locking mechanism is `CompareAndSwap` which means that concurrent writes to the
+same resource always result in the last write winning and potentially losing data. This also allows migrations to
+unknowingly be reverted if an older version of Auth is able to overwrite an already migrated resource.
+
+We need to ensure that:
+
+1. All backend operations are applied in the order they are received, any outdated requests are rejected.
+2. Auth servers know the exact version of a resource requested by clients.
+3. Once a migration has been performed, Auth servers that cannot understand the migrated resource version cannot
+   overwrite the resource.
+4. Migrations are always applied in the correct order and are not skipped.
+5. Migrations can be rolled back without having to manually edit the backend.
+
+#### Resource versioning
+
+The version of a resource MUST be bumped when changes are made to it. Any changes to a resource which can be converted
+into the previous version and do not cause any backward compatibility issues with older Teleport instances only need to
+bump the minor version of a resource. All changes to a resource which would cause backward incompatibility with other
+Auth servers trying to read/write the resource MUST update the major version. All changes to a resource which alter how
+that resource is understood, interpreted, and acted upon by Teleport instances MUST update the major version.
+
+For example, if we originally have a resource like the following at v1:
+
+```proto
+syntax = "proto3";
+
+message Foo {
+  int32 bar = 1;
+}
+```
+
+Then adding a new optional field which was handled appropriately, and defaulted to the correct value by application code
+if empty only requires bumping the version to v1.1 and not require a direct migration.
+
+```diff
+message Foo {
+  int32 bar = 1;
++  string baz = 2;
+}
+```
+
+However, if we were to convert `baz` from a string to `Baz`, that change would not be easily converted into the shape of
+`Foo` at v1.1 and would render the new resource unusable by clients that are only aware of `Foo` v1.1. So, the version
+of the resource must be bumped from v1.1 to v2 to indicate to clients that a breaking change occurred.
+
+```diff
+message Foo {
```

Comment:
Does this mean we will stop putting a version in the type itself (ie `RoleV6` just becomes `Role` and we rely on the version field only)?